presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,552 @@
+        private final int bucketId;
+        private final ValueNode valueNode;
+        private final long valueRawHash;
+        private final long valueOnlyRawHash;
+        private final int nodePointerToUse;
+        private final boolean isEmpty;
+
+        /**
+         * @param currentBuckets
+         * @param bucketId - index into the bucket array. It is writable. If ! EMPTY_BUCKET, it's value points to valid counts and pointers in
+         * @param valueRawHash
+         * @param valueOnlyRawHash
+         * @param nodePointerToUse
+         * @param isEmpty
+         */
+        private BucketDataNode(int bucketId, ValueNode valueNode, long valueRawHash, long valueOnlyRawHash, int nodePointerToUse, boolean isEmpty)
+        {
+            this.bucketId = bucketId;
+            this.valueNode = valueNode;
+            this.valueRawHash = valueRawHash;
+            this.valueOnlyRawHash = valueOnlyRawHash;
+            this.nodePointerToUse = nodePointerToUse;
+            this.isEmpty = isEmpty;
+        }
+
+        private boolean isEmpty()
+        {
+            return isEmpty;
+        }
+
+        /**
+         * true iff needs to update nextNodePoint
+         *
+         * @param groupId
+         * @param block
+         * @param position
+         * @param count
+         * @return
+         */
+        private boolean processEntry(long groupId, Block block, int position, long count)
+        {
+            if (isEmpty()) {
+                addNewGroup(groupId, block, position, count);
+                return true;
+            }
+            else {
+                valueNode.add(count);
+                return false;
+            }
+        }
+
+        private void addNewGroup(long groupId, Block block, int position, long count)
+        {
+            checkState(isEmpty(), "bucket %s not empty, points to %s", bucketId, buckets.get(bucketId));
+
+            // we've already computed the value hash for only the value only; ValueStore will save it for future use
+            int nextValuePosition = valueStore.getPosition(type, block, position, valueOnlyRawHash);
+            valuePositions.set(nodePointerToUse, nextValuePosition);
+            // save hashes for future rehashing
+            valueRawHashList.set(nodePointerToUse, valueRawHash);
+            // set bucket
+            buckets.set(bucketId, nodePointerToUse);
+            // save data
+            counts.set(nodePointerToUse, count);
+            // used for doing value comparisons on hash collisions
+            groupIds.set(nodePointerToUse, groupId);
+            // we only ever store ints as values; we need long as an index
+            int currentHead = (int) headPointers.get(groupId);
+
+            headPointers.set(groupId, nodePointerToUse);
+            nextPointers.set(nodePointerToUse, currentHead);
+        }
+    }
+
+    private class BucketNodeFactory
```

Comment:
I don't think having a class for this buys you anything. I'd make them all (possibly static) methods of the enclosing class.

Reply:
k, works for me
i probably anticipated more functionality and never did a cleanup pass

Reply:
hmm, it does have some helper methods that i'll have to move as well. you don't think it'll clutter up the BucketNode class to have the static create + 2 helper methods?

Reply:
as noted, this was not trivial and required manual work due to classes being created that access
1. members of their own
2. the enclosing class members

i've manually unrolled and will test, see above comment on plan (proposing non-blocking given JMH and JFR results)